sdk/identity/identity/src/msal/nodeFlows/msalClient.ts

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import * as msal from "@azure/msal-node";
+
+import { AccessToken, GetTokenOptions } from "@azure/core-auth";
+import { PluginConfiguration, generatePluginConfiguration } from "./msalPlugins";
+import { credentialLogger, formatSuccess } from "../../util/logging";
+import {
+  defaultLoggerCallback,
+  ensureValidMsalToken,
+  getAuthority,
+  getKnownAuthorities,
+  getMSALLogLevel,
+  publicToMsal,
+} from "../utils";
+
+import { AuthenticationRequiredError } from "../../errors";
+import { IdentityClient } from "../../client/identityClient";
+import { MsalNodeOptions } from "./msalNodeCommon";
+import { getLogLevel } from "@azure/logger";
+import { resolveTenantId } from "../../util/tenantIdUtils";
+
+/**
+ * The logger for all MsalClient instances.
+ */
+const msalLogger = credentialLogger("MsalClient");
+
+/**
+ * Interface for the MSAL (Microsoft Authentication Library) client.
+ * This client is used to interact with Microsoft's identity platform.
+ */
+export interface MsalClient {
+  /**
+   * Retrieves an access token by using a client secret.
+   *
+   * @param scopes - The scopes for which the access token is requested. These represent the resources that the application wants to access.
+   * @param clientSecret - The client secret of the application. This is a credential that the application can use to authenticate itself.
+   * @param options - Additional options that may be provided to the method.
+   * @returns An access token
+   */
+  getTokenByClientSecret(
+    scopes: string[],
+    clientSecret: string,
+    options?: GetTokenOptions,
+  ): Promise<AccessToken>;
+}
+
+/**
+ * Options for creating an instance of the MsalClient.
+ */
+export type CreateMsalClientOptions = Partial<Omit<MsalNodeOptions, "clientId" | "tenantId">>;
+
+/**
+ * Generates the configuration for MSAL (Microsoft Authentication Library).
+ *
+ * @param clientId - The client ID of the application.
+ * @param  tenantId - The tenant ID of the Azure Active Directory.
+ * @param  createMsalClientOptions - Optional. Additional options for creating the MSAL client.
+ * @returns  The MSAL configuration object.
+ */
+export function generateMsalConfiguration(
+  clientId: string,
```

Comment:
Can the env vars be read by the credential itself so that we can keep msalClient params as required? For the credentials based on env vars, these are read by the env var and the config is set by these variables.

Reply:
Ya I was going back and forth. On the one hand, let's say clientId can come from a parameter or from AZURE_CLIENT_ID env var, it doesn't seem great to have each credential do the same logic of `clientId = options.clientId ?? env.AZURE_CLIENT_ID` followed by `if (!clientId) throw new Error()`

It feels like something worth DRYing up - but I also want to keep the MSAL client code as simple as possible... not sure what the right balance is here, but I'll make an issue for us to consider a strategy here. 

It may be as simple as a helper function that is called from here but open to any suggestions / thoughts. The nice thing is we don't have to think about it right now 😄 I can make a tracking issue

Reply:
Not sure if it is helpful here, but in .NET we did this by making the environment the [default value of the options property](https://github.com/Azure/azure-sdk-for-net/blob/7bab1df76fddd6e3981b9ca010882c4436f51ab8/sdk/identity/Azure.Identity/src/Credentials/DefaultAzureCredentialOptions.cs#L196)